hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveSchemaUtils.java

```diff
@@ -0,0 +1,691 @@
+      default:
+        throw new IllegalArgumentException("Unsupported Schema type: " + newFieldSchema.getType().toString());
+    }
+
+    if (evolvedInnerSchema == null) {
+      throw new RuntimeException("Fail to generate evolved schema");
+    }
+
+    return evolvedInnerSchema;
+  }
+
+  private static Schema handleArrayAndMapInnerRecordSchema(Schema oldInnerSchema, Schema newInnerSchema) {
+    Schema evolvedInnerSchema;
+    if (oldInnerSchema.getType().equals(Schema.Type.RECORD)) {
+      evolvedInnerSchema = mergeRecordSchema(oldInnerSchema, newInnerSchema);
+    } else if (oldInnerSchema.getType().equals(Schema.Type.UNION)) {
+      Schema recordTypeSchema = extractTargetTypeFromNullableUnionType(oldInnerSchema, Schema.Type.RECORD);
+      if (recordTypeSchema == null) {
+        throw new IllegalArgumentException("New element Schema is RECORD type, " +
+            "while old element Schema is UNION type without RECORD type in it");
+      }
+
+      Schema evolvedElementSchemaNonnullable = mergeRecordSchema(recordTypeSchema, newInnerSchema);
+      evolvedInnerSchema = createNullableUnionSchema(evolvedElementSchemaNonnullable);
+    } else {
+      throw new IllegalArgumentException(
+          "New element Schema is RECORD type, " + "while old element Schema is neither RECORD type nor UNION type");
+    }
+    return evolvedInnerSchema;
+  }
+
+  private static void validateInputRecordSchema(Schema oldSchema, Schema newSchema) {
+    if (oldSchema == null || newSchema == null) {
+      throw new IllegalArgumentException("The input schemas cannot be null");
+    }
+
+    if (!oldSchema.getType().equals(Schema.Type.RECORD) || !newSchema.getType().equals(Schema.Type.RECORD)) {
+      throw new IllegalArgumentException("Input schemas must be of RECORD type. " + "Old schema type: " +
+          oldSchema.getType() + ", " + "new schema type: " + newSchema.getType());
+    }
+  }
+
+  private static Schema extractTargetTypeFromNullableUnionType(Schema unionSchema, Schema.Type targetType) {
+    if (unionSchema == null ||
+        !unionSchema.getType().equals(Schema.Type.UNION) ||
+        !AvroSchemaUtil.isOptionSchema(unionSchema)) {
+      throw new IllegalArgumentException("Input schemas must be of nullable UNION type.");
+    }
+
+    for (Schema field : unionSchema.getTypes()) {
+      if (field.getType().equals(targetType)) {
+        return field;
+      }
+    }
+
+    return null;
+  }
+
+  private static Schema createNullableUnionSchema(Schema schema) {
+    if (schema == null) {
+      throw new IllegalArgumentException("The input schema cannot be null");
+    }
+
+    List<Schema> unionSchemas = new ArrayList<>();
+    unionSchemas.add(Schema.create(Schema.Type.NULL));
+    unionSchemas.add(schema);
+
+    return Schema.createUnion(unionSchemas);
+  }
+
+  static boolean isRecordSchemaEvolved(Schema oldSchema, Schema newSchema) {
```

Comment:
Why do we have two methods with slightly different names `isRecordSchemaEvolved` and `isSchemaEvolved`?